common/src/types/v0/store/node.rs

```diff
@@ -66,33 +140,237 @@ impl NodeSpec {
+        &self.cordon_drain_state
+    }
+
     /// Node gRPC endpoint
     pub fn set_endpoint(&mut self, endpoint: String) {
         self.endpoint = endpoint
     }
+
     /// Cordon node by applying the label.
     pub fn cordon(&mut self, label: String) {
-        self.cordon_labels.push(label);
+        // the the node has the label, return with an error
+        match &mut self.cordon_drain_state {
+            Some(ds) => match ds {
+                CordonDrainState::Cordoned(state) => {
+                    // just add the label
+                    state.add_cordon_label(label);
+                }
+                CordonDrainState::Draining(state) => {
+                    // just add the label
+                    state.add_cordon_label(label);
+                }
+                CordonDrainState::Drained(state) => {
+                    // just add the label
+                    state.add_cordon_label(label);
+                }
+            },
+            None => {
+                //add the label and set the state to cordoned
+                self.cordon_drain_state =
+                    Some(CordonDrainState::Cordoned(CordonedState::new(vec![label])));
+            }
+        }
+    }
+
+    /// Drain node by applying the drain label.
+    pub fn drain(&mut self, label: String) {
+        // the the node has the label, return with an error
+        match &mut self.cordon_drain_state {
+            Some(ds) => match ds {
+                CordonDrainState::Cordoned(state) => {
+                    // set state to draining and add label
+                    let drainlabels = vec![label];
+                    self.cordon_drain_state = Some(CordonDrainState::Draining(DrainState::new(
+                        state.cordonlabels.clone(),
+                        drainlabels,
+                    )));
+                }
+                CordonDrainState::Draining(state) => {
+                    // add the label
+                    state.add_drain_label(label);
+                }
+                CordonDrainState::Drained(state) => {
+                    // add the label
+                    state.add_drain_label(label);
+                }
+            },
+            None => {
+                //add the label and set the state to draining
+                let cordonlabels: Vec<String> = vec![];
+                let drainlabels = vec![label];
+                self.cordon_drain_state = Some(CordonDrainState::Draining(DrainState::new(
+                    cordonlabels,
+                    drainlabels,
+                )));
+            }
+        }
     }
     /// Uncordon node by removing the corresponding label.
     pub fn uncordon(&mut self, label: String) {
-        if let Some(index) = self.cordon_labels.iter().position(|l| l == &label) {
-            self.cordon_labels.remove(index);
+        // if the label is a cordon label
+        if self.has_cordon_only_label(label.clone()) {
```

Comment:
`has_cordon_only_label` can take a `&str` to avoid cloning.

Reply:
Do we need to check at all though? Can't we just do
```rust
match &mut self.cordon_drain_state {
Some(ds) => match ds {
    CordonDrainState::Cordoned(state) => {
        state.remove_label(label);
    }
    CordonDrainState::Draining(state) => {
        state.remove_label(label);
        ...
    }
    CordonDrainState::Drained(state) => {
        state.remove_label(label);
        ...
    }
},
None => {
   // Not cordoned or drained
}
```

Reply:
Yes, that works 👍